Add permission node for using jukeboxes, fix worldguard integration

neomoth opened the pull request:
Addresses the following issue/feature request by adding the requested permission node:
- #285

Addresses the fact that the WorldGuard integration does not account for unset flags that don't have a default permission (CHEST_ACCESS does not have a default value when unset, meaning by default players cannot use jukeboxes.)
This is done via altering the flag check to allow returning `null`, and explicitly check that the returned result is equal to `false`.

neomoth commented:
May need some slight tweaking, unsure how you want to handle this, and the CodeFactor thing is pointing out some stuff but also I barely know how Kotlin works in the first place. Just wanted to address these issues as they were causing issues for the server I manage.

spartacus04 commented:
Hello! Thanks a lot for the pr. I've been extremely busy lately and I've had no time to work on the plugin. It's extremely appreciated.

I'll just check the code a bit in a few minutes, and tell you if there are any changes you should make.

spartacus04 commented on src/main/resources/plugin.yml:
This should be `true` and not `op`, because by default we want all users to use the jukeboxes

neomoth replied:
Makes sense yeah

neomoth replied:
Updated plugin.yml

spartacus04 replied:
Thanks!

spartacus04 commented on src/main/java/me/spartacus04/jext/listeners/JukeboxClickEvent.kt:
```diff
@@ -31,18 +31,10 @@ internal class JukeboxClickEvent : JextListener() {
-            event.player.sendMessage("Missing permission")
             event.isCancelled = true
             return
         }
```
This check should be moved to the `OnJukeboxInteract` method, you can also remove the `|| event.player.isOp` since it should be managed by a permission manager and not the plugin

spartacus04 commented on src/main/java/me/spartacus04/jext/listeners/JukeboxClickEvent.kt:
```diff
@@ -67,11 +59,9 @@ internal class JukeboxClickEvent : JextListener() {
         if(!(event.player.hasPermission("jext.usejukebox")||event.player.isOp)){
-            event.player.sendMessage("Missing permission")
             return
         }
```
Just like before, this check should be moved to the OnJukeboxInteract method, you can also remove the || event.player.isOp since it should be managed by a permission manager and not the plugin

spartacus04 commented on src/main/java/me/spartacus04/jext/listeners/JukeboxClickEvent.kt:
```diff
@@ -30,9 +30,13 @@ internal class JukeboxClickEvent : JextListener() {
-        if(!INTEGRATIONS.hasJukeboxAccess(event.player, block)) {
-             event.isCancelled = true
-             return
+        if(INTEGRATIONS.hasJukeboxAccess(event.player, block)==false) {
+            event.isCancelled = true
+            return
+        }
```
Just use the `!` before the check, this may not make sense now because you made the function nullable but we'll make it not nullable in the next few functions

spartacus04 commented on src/main/java/me/spartacus04/jext/listeners/JukeboxClickEvent.kt:
```diff
@@ -54,8 +58,12 @@ internal class JukeboxClickEvent : JextListener() {
         event.isCancelled = true
-
-        if(!INTEGRATIONS.hasJukeboxGuiAccess(event.player, block)) return
+        if(INTEGRATIONS.hasJukeboxGuiAccess(event.player, block)==false) {
+            return
+        }
```
Just use the ! before the check, this may not make sense now because you made the function nullable but we'll make it not nullable in the next few functions

spartacus04 commented on src/main/java/me/spartacus04/jext/integrations/WorldGuardIntegration.kt:
```diff
@@ -15,19 +15,25 @@ internal class WorldGuardIntegration : Integration {
-    override fun hasJukeboxAccess(player: Player, block: Block): Boolean = canInteract(player, block, Flags.CHEST_ACCESS)
+    override fun hasJukeboxAccess(player: Player, block: Block): Boolean? = canInteract(player, block, Flags.CHEST_ACCESS)
 
-    override fun hasJukeboxGuiAccess(player: Player, block: Block): Boolean = canInteract(player, block, Flags.BUILD, Flags.INTERACT, Flags.USE)
+    override fun hasJukeboxGuiAccess(player: Player, block: Block): Boolean? = canInteract(player, block, Flags.BUILD, Flags.INTERACT, Flags.USE)
 
-    private fun canInteract(player: Player, block: Block, vararg flags: StateFlag) : Boolean {
+    private fun canInteract(player: Player, block: Block, vararg flags: StateFlag) : Boolean? {
```
Make the methods not nullable

spartacus04 commented on src/main/java/me/spartacus04/jext/integrations/WorldGuardIntegration.kt:
```diff
@@ -15,19 +15,25 @@ internal class WorldGuardIntegration : Integration {
 
-        return flags.any {
-            containerQuery.testState(BukkitAdapter.adapt(block.location), wgPlayer, it)
+        for(flag in flags) {
+            val state = containerQuery.queryState(BukkitAdapter.adapt(block.location), wgPlayer, flag)
+            return when(state) {
+                StateFlag.State.ALLOW -> true
+                StateFlag.State.DENY -> continue
+                null -> null
+            }
         }
+        return false
```
You can use the `flags.any` method and do the null check inside by doing

```
flags.any {
    containerQuery.testState(BukkitAdapter.adapt(block.location), wgPlayer, it) == StateFlag.State.ALLOW
}
```

spartacus04 commented on src/main/java/me/spartacus04/jext/integrations/IntegrationsManager.kt:
```diff
@@ -19,9 +20,29 @@ class IntegrationsManager {
 
-    fun hasJukeboxAccess(player: Player, block: Block): Boolean = integrations.all { it.hasJukeboxAccess(player, block) }
+    fun hasJukeboxAccess(player: Player, block: Block): Boolean? {
+        for(integration in integrations){
+            val state = integration.hasJukeboxAccess(player, block)
+            return when(state){
+                true->true
+                false->continue
+                null->null
+            }
+        }
+        return false
+    }
 
-    fun hasJukeboxGuiAccess(player: Player, block: Block): Boolean = integrations.all { it.hasJukeboxGuiAccess(player, block) }
+    fun hasJukeboxGuiAccess(player: Player, block: Block): Boolean? {
+        for(integration in integrations){
+            val state = integration.hasJukeboxGuiAccess(player, block)
+            return when(state){
+                true->true
+                false->continue
+                null->null
+            }
+        }
+        return false
+    }
```
Since the methods are not nullable you can do the old `integrations.all` function

spartacus04 commented on src/main/java/me/spartacus04/jext/integrations/Integration.kt:
```diff
@@ -18,12 +18,12 @@ interface Integration {
      * @param block The "block" parameter represents the jukebox block that the player is trying to interact with.
      */
-    fun hasJukeboxGuiAccess(player: Player, block: Block): Boolean
+    fun hasJukeboxGuiAccess(player: Player, block: Block): Boolean?
```
Make not nullable

spartacus04 commented on src/main/java/me/spartacus04/jext/integrations/Integration.kt:
```diff
@@ -18,12 +18,12 @@ interface Integration {
      * @param block The "block" parameter represents the jukebox block that the player is trying to interact with.
      */
-    fun hasJukeboxAccess(player: Player, block: Block): Boolean
+    fun hasJukeboxAccess(player: Player, block: Block): Boolean?
```
Make not nullable

spartacus04 reviewed (changes requested):
I've requested a few changes to make the code mode readable and coincise. Since you said that you are new to Kotlin, if you have any doubts or questions don't hesitate to ask

spartacus04 commented on src/main/java/me/spartacus04/jext/listeners/JukeboxClickEvent.kt:
```diff
@@ -30,9 +30,13 @@ internal class JukeboxClickEvent : JextListener() {
+            event.isCancelled = true
+            return
+        }
+        if(!(event.player.hasPermission("jext.usejukebox")||event.player.isOp)){
```
This check should be moved to the OnJukeboxInteract method, you can also remove the `|| event.player.isOp` since it should be managed by a permission manager and not the plugin

spartacus04 commented:
@neomoth If you are busy I can fix these problems myself don't worry, I just reported these to make them clear since it could help you get better at kotlin

neomoth commented:
I've made the requested changes, I had changed so much due to not being familiar with kotlin syntax and wasn't aware I could have avoided the whole needing to check for null situation by just altering the original `flags.any` return statement lol.
Also, I updated on browser due to being a tad busy, so I can't test if this actually *works* though I assume it does and can test in a couple of hours just in case.

neomoth commented:
Oh that's a problem.
```
Operator '==' cannot be applied to 'Boolean' and 'StateFlag.State'
```

spartacus04 commented:
Just replace `containerQuery.testState` with `containerQuery.queryState`

neomoth commented:
Ohhh i see. One moment.

neomoth commented:
Fixed, had to actually rewind and commit again because apparently it really didn't feel like using my correct credentials on this machine so it had used a different name and email for the commit lol. Tested and it works, I'm able to use a jukebox deopped in the `__global__` region.

spartacus04 commented:
Thanks again for the pr!
